stdlib/public/core/Unicode.swift

```diff
@@ -141,83 +99,64 @@ public struct UTF8 : UnicodeCodecType {
-          return false
-        }
-      case 0xed:
-        if cu1 < 0x80 || cu1 > 0x9f {
-          return false
-        }
-      case 0xf0:
-        if cu1 < 0x90 || cu1 > 0xbf {
-          return false
-        }
-      case 0xf4:
-        if cu1 < 0x80 || cu1 > 0x8f {
-          return false
-        }
-      default:
-        _sanityCheck(cu0 >= 0xc2 && cu0 <= 0xf4,
-            "invalid first bytes should be handled in the caller")
-        if cu1 < 0x80 || cu1 > 0xbf {
-          return false
-        }
-      }
-      return true
 
-    default:
-      _sanityCheckFailure("one-byte sequences should be handled in the caller")
-    }
-  }
-
-  /// Returns `true` if the LSB bytes in `buffer` are well-formed UTF-8 code
+  /// Returns `true` if the LSB bytes in `buffer` are well-formed UTF-8 code		
   /// unit sequence.
   @warn_unused_result
   static func _isValidUTF8(buffer: UInt32, validBytes: UInt8) -> Bool {
     _sanityCheck(validBytes & 0b0000_1111 != 0,
         "input buffer should not be empty")
 
-    let cu0 = UInt8(buffer & 0xff)
-    let trailingBytes = _numTrailingBytes(cu0)
-    switch trailingBytes {
-    case 0:
-      return true
-
-    case 1, 2, 3:
-      // We *don't* need to check the if the buffer actually contains at least
-      // `trailingBytes` bytes.  Here's why.
-      //
-      // If the buffer is not full -- contains fewer than 4 bytes, we are at
-      // EOF, and the buffer will be padded with 0x00.  Thus, an incomplete
-      // code unit sequence just before EOF would be seen by code below as
-      // padded with nuls.  This sequence will be rejected by the logic in
-      // `_isValidUTF8Impl`, because the nul byte is not a valid continuation
-      // byte for UTF-8.
-      return _isValidUTF8Impl(buffer, length: trailingBytes + 1)
+    if _fastPath(buffer & 0x80 == 0) {
+      return true // 0x00 -- 0x7f: 1-byte sequences (ASCII).
+    }
 
-    default:
+    // Determine sequence length using high 5 bits of 1st byte. We use a
+    // look-up table to branch less. 1-byte sequences are handled above.
+    //
+    //  case | pattern | description
+    // ----------------------------
+    //   00  |  110xx  | 2-byte sequence
+    //   01  |  1110x  | 3-byte sequence
+    //   10  |  11110  | 4-byte sequence
+    //   11  |  other  | invalid
+    //
+    //                     11xxx      10xxx      01xxx      00xxx
+    let lut0: UInt32 = 0b1011_0000__1111_1111__1111_1111__1111_1111
+    let lut1: UInt32 = 0b1100_0000__1111_1111__1111_1111__1111_1111
+
+    let index = (buffer >> 3) & 0x1f
+    let bit0 = (lut0 >> index) & 1
+    let bit1 = (lut1 >> index) & 1
+
+    // Note that all past-EOF bytes are 0 and thus never a valid continuation.
+    switch (bit1, bit0) {
+    case (0, 0): // 2-byte sequence.
+      // Require 10xx xxxx  110x xxxx.
+      if buffer & 0xc0e0 != 0x80c0 { return false }
+      // Disallow xxxx xxxx  xxx0 000x (<= 7 bits case).
+      if buffer & 0x001e == 0x0000 { return false }
```

Comment:
Sorry, I don't understand this case.  I think you meant to test against `0x1f00` instead of `0x001e`.